ccl/backupccl: TestRestoreAsOfSystemTimeGCBounds failed

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/e28999f8-4d51-4e92-b3d9-a9e402595828?testReportRun=30&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [fce4d4723519bc4ca6e9ef5da0ae19960c84752c](https://github.com/cockroachdb/cockroach/commits/fce4d4723519bc4ca6e9ef5da0ae19960c84752c):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m55s)

```
Stack: 

```
goroutine 559425 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds3423607260
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=30</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35813

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/9bf21f38-47fd-44ee-9ab1-67dabe38b27d?testReportRun=27&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [24547fb1115903d432004207fc49a87262133c34](https://github.com/cockroachdb/cockroach/commits/24547fb1115903d432004207fc49a87262133c34):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m55s)

```
Stack: 

```
goroutine 548950 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds502525171
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=27</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/f516e900-d9ad-494d-b367-e793e7896d08?testReportRun=14&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [a60fc9aa732dc25a36d14415cee29c41a59b48a7](https://github.com/cockroachdb/cockroach/commits/a60fc9aa732dc25a36d14415cee29c41a59b48a7):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m56s)

```
Stack: 

```
goroutine 477873 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds1473938618
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=14</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/1c5d467f-bca9-4455-bf34-a20a843efb1f?testReportRun=27&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [2420e5c067c23eb5323f298a5f11d6d86d1d45b1](https://github.com/cockroachdb/cockroach/commits/2420e5c067c23eb5323f298a5f11d6d86d1d45b1):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m56s)

```
Stack: 

```
goroutine 465650 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds2992306226
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=27</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/1c82e2bc-758a-44a2-a100-b185ad2f686d?testReportRun=14&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [b2e31876366324c2ebe5c2ad8bbd644997e90864](https://github.com/cockroachdb/cockroach/commits/b2e31876366324c2ebe5c2ad8bbd644997e90864):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m55s)

```
Stack: 

```
goroutine 476865 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds3546588989
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=14</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/backupccl.TestRestoreAsOfSystemTimeGCBounds [failed](https://tanzanite.cluster.engflow.com/invocations/default/11ef7a11-4af0-4e81-aadd-d223e9d62a50?testReportRun=29&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [b2e31876366324c2ebe5c2ad8bbd644997e90864](https://github.com/cockroachdb/cockroach/commits/b2e31876366324c2ebe5c2ad8bbd644997e90864):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestRestoreAsOfSystemTimeGCBounds (14m55s)

```
Stack: 

```
goroutine 478649 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRestoreAsOfSystemTimeGCBounds
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRestoreAsOfSystemTimeGCBounds3310012189
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:157: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=29</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreAsOfSystemTimeGCBounds.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -3911,11 +3911,7 @@ func TestRestoreAsOfSystemTimeGCBounds(t *testing.T) {
 
 	const numAccounts = 10
 	ctx := context.Background()
-	args := base.TestClusterArgs{
-		ServerArgs: base.TestServerArgs{
-			DefaultTestTenant: base.TestDoesNotWorkWithSecondaryTenantsButWeDontKnowWhyYet(118625),
-		},
-	}
+	args := base.TestClusterArgs{}
 	tc, sqlDB, _, cleanupFn := backupRestoreTestSetupWithParams(t, singleNode, numAccounts, InitManualReplication, args)
 	defer cleanupFn()
 	const dir = "nodelocal://1/"

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -915,6 +915,8 @@ type Manager struct {
 	descUpdateCh chan catalog.Descriptor
 	// descDelCh receives deleted descriptors from the range feed.
 	descDelCh chan descpb.ID
+	// rangefeedErrCh receives any terminal errors from the rangefeed.
+	rangefeedErrCh chan error
 }
 
 const leaseConcurrencyLimit = 5
@@ -1006,6 +1008,7 @@ func NewLeaseManager(
 	lm.draining.Store(false)
 	lm.descUpdateCh = make(chan catalog.Descriptor)
 	lm.descDelCh = make(chan descpb.ID)
+	lm.rangefeedErrCh = make(chan error)
 	return lm
 }
 
@@ -1437,7 +1440,6 @@ func (m *Manager) RefreshLeases(ctx context.Context, s *stop.Stopper, db *kv.DB)
 				if evFunc := m.testingKnobs.TestingDescriptorRefreshedEvent; evFunc != nil {
 					evFunc(desc.DescriptorProto())
 				}
-
 			case <-s.ShouldQuiesce():
 				return
 			}
@@ -1513,6 +1515,13 @@ func (m *Manager) watchForUpdates(ctx context.Context) {
 		ctx, "lease", []roachpb.Span{descriptorTableSpan}, hlc.Timestamp{}, handleEvent,
 		rangefeed.WithSystemTablePriority(),
 		rangefeed.WithOnCheckpoint(handleCheckpoint),
+		rangefeed.WithOnInternalError(func(ctx context.Context, err error) {
+			select {
+			case m.rangefeedErrCh <- err:
+			case <-ctx.Done():
+			case <-m.stopper.ShouldQuiesce():
+			}
+		}),
 	)
 }
 
@@ -1610,6 +1619,10 @@ func (m *Manager) RunBackgroundLeasingTask(ctx context.Context) {
 				rangeFeedProgressWatchDogTimeout,
 					rangeFeedProgressWatchDogEnabled = m.getRangeFeedMonitorSettings()
 				rangeFeedProgressWatchDog.Reset(rangeFeedProgressWatchDogTimeout)
+			case err := <-m.rangefeedErrCh:
+				log.Warningf(ctx, "lease rangefeed failed with error: %s", err.Error())
+				m.handleRangeFeedError(ctx)
+				m.refreshSomeLeases(ctx, true /*refreshAll*/)
 			case <-refreshTimer.C:
 				refreshTimer.Read = true
 				refreshTimer.Reset(m.storage.jitteredLeaseDuration() / 2)
@@ -1655,12 +1668,23 @@ func (m *Manager) handleRangeFeedAvailability(ctx context.Context) {
 	// the range feed.
 	if timeutil.Since(m.mu.rangeFeedIsUnavailableAt) >=
 		LeaseMonitorRangeFeedResetTime.Get(&m.settings.SV) {
-		log.Warning(ctx, "attempting restart of leasing range feed")
-		// Attempt a range feed restart if it has been down too long.
-		m.watchForUpdates(ctx)
-		// Track when the last restart occurred.
-		m.mu.rangeFeedIsUnavailableAt = timeutil.Now()
+		m.restartLeasingRangeFeedLocked(ctx)
 	}
+}
+
+func (m *Manager) handleRangeFeedError(ctx context.Context) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.restartLeasingRangeFeedLocked(ctx)
+}
+
+func (m *Manager) restartLeasingRangeFeedLocked(ctx context.Context) {
+	log.Warning(ctx, "attempting restart of leasing range feed")
+	// Attempt a range feed restart if it has been down too long.
+	m.watchForUpdates(ctx)
+	// Track when the last restart occurred.
+	m.mu.rangeFeedIsUnavailableAt = timeutil.Now()
+	m.mu.rangeFeedCheckpoints = 0
 }
 
 // cleanupExpiredSessionLeases expires session based leases marked for removal,
@@ -1818,7 +1842,7 @@ func (m *Manager) DeleteOrphanedLeases(ctx context.Context, timeThreshold int64)
 		// Read orphaned leases, and join against the internal session
 		// table in case we have dual written leases.
 		query := `
-SELECT COALESCE(l."descID", s."desc_id") as "descID", COALESCE(l.version, s.version), l.expiration, s."session_id", l.crdb_region, s.crdb_region FROM 
+SELECT COALESCE(l."descID", s."desc_id") as "descID", COALESCE(l.version, s.version), l.expiration, s."session_id", l.crdb_region, s.crdb_region FROM
 	 system.public.lease as l FULL OUTER JOIN "".crdb_internal.kv_session_based_leases as s ON l."nodeID"=s."sql_instance_id" AND
 	  l."descID"=s."desc_id" AND l.version=s.version
 		WHERE COALESCE(l."nodeID", s."sql_instance_id") =%d

